### ISSUE TITLE
Crash in didRegisterForRemoteNotifications

### ISSUE THREAD

**Issue by skuske** [OneSignal/OneSignal-iOS-SDK#513]
We are seeing this issue when calling `didRegisterForRemoteNotifications`

`+[NSString hexStringFromData:]: unrecognized selector sent to class 0x1cb196628`

Crashlytics detected the issue in

`+[OneSignal didRegisterForRemoteNotifications:deviceToken:] + 2092 (OneSignal.m:2092)`

Tested on iOS 12.4

Any idea? TIA.

**Comment by Nightsd01** [comment]
@skuske this would be related to a recent change we made for iOS 13 compatibility. For the time being to unblock development I’d recommend using the previous version. 

Can you provide a few details:
- Is this crash occurring for you 100% of the time?
- Is it crashing on multiple devices or just one?
- How did you install the SDK (cocoapods, carthage, manual)?

**Comment by skuske** [comment]
@Nightsd01 

Thanks for your reply. 

1.) Yes, it crashes all the time
2.) Yes, it crashes on multiple devices
3.) I installed manually

**Comment by jkasten2** [comment]
@skuske Can you share a sample project reproducing the issue?

I tried the latest 2.10.1 release on our [example project](https://github.com/OneSignal/OneSignal-iOS-SDK/tree/master/Examples/ObjectiveCExample) and was not able to reproduce the issue.

**Comment by skuske** [comment]
@jkasten2 

I set up a test project with exactly the same 

`OneSignal initWithLaunchOptions`

call, but the crash does not happen there. I even re-coded the whole OneSignal integration within the crashing app, followed the setup guide, but no luck. Still the same 

`+[NSString hexStringFromData:]: unrecognized selector sent to class 0x1cb196628`

crash in 2.10.1.

No crash in 2.10.0.

**Comment by Nightsd01** [comment]
@skuske this doesn’t seem to be a common issue so if you can please reproduce it in the demo project and create a .zip and post the link here that would be very helpful - I’ll investigate the issue on Saturday

**Comment by skuske** [comment]
@jkasten2 

As I mentioned above, the problem does not occur in a test project we created. The problem does not occur in v2.10.0, so it has to be related to a change in v2.10.1.

**Comment by Nightsd01** [comment]
@skuske I tested this the other day and couldn’t reproduce. Until you can send us the demo project in a .zip there’s not much we can really do. This is the third time we have asked

**Comment by jkasten2** [comment]
@skuske This is an odd problem of why the selector would not be found. The selector `+[NSString hexStringFromData:]` was adding in 2.10.1 so this error would have not shown up in earlier versions.

There are two things that come to mind of what could be causing the issue;
1. An Xcode settings that is affecting how classes are being loaded
2. Another SDK or library that is inverfering in some way

@skuske Can you try copying in the same dependencies and Xcode setting from your project to try to reproduce the problem?
Another option is if you don't mind sharing your project with OneSignal support then they could help take a look. You could even delete your projects code if that is an issue before sending.
Thanks.

**Comment by skuske** [comment]
@Nightsd01 

I understand that you asked for a sample project, and I have already mentioned twice that I myself can not reproduce the problem in a sample project. Even so, version 2.10.0 works fine, but version 2.10.1 unfortunately does not produce the mentioned bug in a sample project. So the problem lies in version 2.10.1, even if I can not reproduce it in a sample project, because the converse conclusion, which results from the error-free use of 2.10.0, shows me that 2.10.1 is not faultless.

@JKalash 

Once I find the time I'll try to set up a sample project with all the same dependencies as in the original (crashing) project.

**Comment by skuske** [comment]
@Nightsd01

I understand that you asked for a sample project, and I have already mentioned twice that I myself can not reproduce the problem in a sample project. Even so, version 2.10.0 works fine, but version 2.10.1 unfortunately does not produce the mentioned bug in a sample project. So the problem lies in version 2.10.1, even if I can not reproduce it in a sample project, because the converse conclusion, which results from the error-free use of 2.10.0, shows me that 2.10.1 is not faultless.

@jkasten2 

Once I find the time I'll try to set up a sample project with all the same dependencies as in the original (crashing) project. Another idea: could you compile a modified framework with  `+[NSString hexStringFromData:]` renamed to something else, so that I can test if this is indeed a conflict with another dependency using a method with the same name?

**Comment by jkasten2** [comment]
@skuske Your error is pointing to the selector not being found so I don't believe it is a conflict. It shows `unrecognized selector sent to class 0x1cb196628` so the issue might be that the category didn't get attached to `NSString` or the instance was not a `NSString`.

**Comment by skuske** [comment]
@jkasten2 

Got it. I cleaned up the project, deleted the DerivedData folder and now it seems to work. Thanks for your help! :o) I'll do some more tests tomorrow and will then close the issue if it continues to work.

**Comment by skuske** [comment]
@jkasten2 

No, the issue is back. Even after cleaning the project and deleting the DerivedData folder, it crashes again with the same exception

+[NSString hexStringFromData:]: unrecognized selector sent to class 0x1cb196628

I think I give up and will use one of the previous versions which seem to work fine. Not sure if that is the case on iOS 13, though...

**Comment by skuske** [comment]
I did some more tests: running our project with 2.10.0 on iOS 12 works fine, and the device gets registered with OneSignal. I then replaced the framework with 2.10.1 and it works **without** crashing.

However, after deleting the app and running it again on the same iOS 12 device with framework 2.10.1 causes the crash again. So, it indeed looks like it is only happening on the initial registration with OneSignal, and only on iOS 12 devices. If the device was registered before, then 2.10.1 works fine.

I cannot test that on iOS 11 or below. Don't have any such devices here.

**Comment by jkasten2** [comment]
@skuske I tried to reproduce the issue with iOS 12.3 and iOS 9.3.6 with both Xcode 10.3 and Xcode 11 Beta 5. I uninstall the app each time since this is what you said reproduce it but no luck. I was testing on our example projects:
https://github.com/OneSignal/OneSignal-iOS-SDK/tree/master/Examples
Our example project use cocoapods but that most likely won't be a factor.

In order for us to find the root of the issue we will need a way to reproduce the crash. Please provide exact steps to reproduce on a new Xcode project or provide an example project. Thanks.

**Comment by skuske** [comment]
@jkasten2 

Since it is unfortunately impossible for me to deliver a corresponding example project in which the error occurs, we must wait until the problem occurs with someone else.

I can not break up the original project until I'm allowed to send it to you for debugging. We have already spent enough time analyzing this problem and finding the possible cause. Therefore, we will now use v2.10.1, but register with OneSignal only on iOS 13 devices. This solves the problem, at least for the time being, and since sooner or later, approximately 90-95% of all our users will be upgrading to iOS 13, this is the most viable solution.

Thanks.

**Comment by skuske** [comment]
@jkasten2 

I compiled the framework myself and added it to the crashing project. What I have found out now is that 

`inDeviceToken`

is `nil` in line 2092 of OneSignal.m

That's why it crashes. Any idea why it might be `nil` and how to fix that? Anything I can try or debug?

**Comment by Zorrendor** [comment]
I have the same error building with Unity the latest OneSignal version.
XCode Version 10.2.1
iPhone6 iOS 12.3

**Comment by skuske** [comment]
@Zorrendor 

Glad to hear that finally it is not just me seeing this. 

The only way to get around that at the moment is by either not supporting iOS 13 (then use framework 2.10.0), or if you need to support iOS 13, then use 2.10.1, but do then **not** call `OneSignal initWithLaunchOptions` on devices with iOS 12 and below, as that will crash.

As I mentioned above, most people will upgrade to iOS 13 sooner or later anyway, which means that once they upgraded to iOS 13 you can let your app call `OneSignal initWithLaunchOptions`

**Comment by jkasten2** [comment]
@skuske On your screenshot, I have seen the debugger mistakenly show values as `nil` when they are not. I have added some tests in the following PR to ensure `NSString.hexStringFromData` can handle `nil`, `NULL`, and `[NSData new]` to ensure this isn't the issue.
https://github.com/OneSignal/OneSignal-iOS-SDK/pull/516

The error `unrecognized selector` error is saying that `hexStringFromData:` can not be found on `NSString`. This is a selected add by the OneSignal SDK to the `NSString` class using the [Objective-C Category](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html) feature. However it seems the category feature does not work unless `-ObjC` is set on `Build Settings` > `Other Linker Flags` on the project consuming the OneSignal SDK. The iOS version did not seem to have any effect on this.

However I found a way to load all OneSignal classes via a "Perform Single-Object Prelink" that should fix this issue without us know needing remove all categories or require everyone to add `-ObjC`.

I'll be making a PR for this soon and it will in the next release.
